clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java

```diff
@@ -66,19 +65,20 @@ protected AbstractTachyonFileSystem() {
   }
 
   @Override
-  public long create(TachyonURI path, CreateOptions options) throws IOException, TachyonException {
+  public long create(TachyonURI path, CreateOptions options)
+      throws FileAlreadyExistsException, IOException, InvalidPathException, TachyonException {
     FileSystemMasterClient masterClient = mContext.acquireMasterClient();
     try {
-      long fileId =
-          masterClient.createFile(path.getPath(), options.getBlockSize(), options.isRecursive(),
-              options.getTTL());
-      return fileId;
-    } catch (BlockInfoException e) {
-      throw new TachyonException(e, TachyonExceptionType.FILE_ALREADY_EXISTS);
-    } catch (FileAlreadyExistException e) {
-      throw new TachyonException(e, TachyonExceptionType.FILE_ALREADY_EXISTS);
-    } catch (InvalidPathException e) {
-      throw new TachyonException(e, TachyonExceptionType.INVALID_PATH);
+      return masterClient.createFile(path.getPath(), options.getBlockSize(), options.isRecursive(),
+          options.getTTL());
+    } catch (TachyonException e) {
+      if (e.getType() == TachyonExceptionType.BLOCK_INFO) {
+        throw new FileAlreadyExistsException(e.getMessage(), e);
```

Comment:
The thrift exception is converted to a `TachyonException` at the `masterClient` level, but we unwrap specific Tachyon exceptions at the client boundary. This way, Thrift exceptions are still isolated from the entire codebase, but if we change the exception specification, only the client code will have to change, rather than the entire codebase.